### ISSUE TITLE
Nightly build errors, intel/2023.2.0 (both icpc and icpx) with c++20 enabled: MDSpan/Kokkos_MDSpan_Header.hpp(27): catastrophic error: cannot open source file...

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#7075]
**Describe the bug**

Following merge of #7069 to enable mdspan by default, builds with intel/2023.2.0 (both icpc and icpx) and c++20 support enabled fail to compile with error:
```
17:27:37 cd /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/Build/core/src && /projects/x86-64-icelake-rocky8/compilers/intel-oneapi-compilers/2023.2.0/gcc/8.5.0/base/2lqdmcu/compiler/2023.2.0/linux/bin/intel64/icpc -DKOKKOS_DEPENDENCE -I/home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/Build/core/src -I/home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src -I/home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/Build -I/home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/../../tpls/desul/include -I/home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/../../tpls/mdspan/include -Wall -O3 -DNDEBUG -std=c++20 -qopenmp -MD -MT core/src/CMakeFiles/kokkoscore.dir/OpenMP/Kokkos_OpenMP_Instance.cpp.o -MF CMakeFiles/kokkoscore.dir/OpenMP/Kokkos_OpenMP_Instance.cpp.o.d -o CMakeFiles/kokkoscore.dir/OpenMP/Kokkos_OpenMP_Instance.cpp.o -c /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/OpenMP/Kokkos_OpenMP_Instance.cpp
17:27:37 In file included from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Extents.hpp(25),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/Kokkos_View.hpp(40),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/Kokkos_Parallel.hpp(31),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/OpenMP/Kokkos_OpenMP.hpp(32),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/decl/Kokkos_Declare_OPENMP.hpp(21),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/Build/KokkosCore_Config_DeclareBackend.hpp(22),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/Kokkos_Core.hpp(45),
17:27:37                  from /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/impl/Kokkos_HostSpace_deepcopy.cpp(21):
17:27:37 /home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Header.hpp(27): catastrophic error: cannot open source file "/home/jenkins/blake-new/workspace/Kokkos_develop_blake_intel_2023_2_0-icpc_openmp_serial_cpp20/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Header.hpp"
17:27:37   #include <version>
```

@crtrott can you help advise on how to proceed?

**Please include the following for a minimal reproducer**

**Reproducer**: (blake all queue)
```
source /projects/x86-64-icelake-rocky8/spack-config/blake-setup-user-module-env.sh
module purge
module load intel-oneapi-compilers/2023.2.0 intel-oneapi-dpl/2022.2.0 intel-oneapi-tbb/2021.10.0 cmake git

# icpx configuration
$KOKKOS_PATH/generate_makefile.bash --with-openmp --with-serial --arch=SPR --compiler=icpx --cxxflags="-fp-model=precise -Wall" --cxxstandard=20

# icpc configuration
$KOKKOS_PATH/generate_makefile.bash --with-openmp --with-serial --compiler=icpc --cxxflags="-Wall" --cxxstandard=20
```

**Comment by ndellingwood** [comment]
This is also impacting kokkos-kernels builds with similar configuration, sample output snip:

```
17:35:55 /home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_OneAPI2023_2_0_ICPC_Serial/Trilinos/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Header.hpp(47): catastrophic error: cannot open source file "/home/jenkins/blake-new/workspace/KokkosEco_Trilinos_Blake_OneAPI2023_2_0_ICPC_Serial/Trilinos/kokkos/core/src/View/MDSpan/Kokkos_MDSpan_Header.hpp"
17:35:55   #include <mdspan/mdspan.hpp>
```

**Comment by ndellingwood** [comment]
I can temporarily disable mdspan by default in the Trilinos integration jobs to keep coverage until the issue is resolved

**Comment by ndellingwood** [comment]
In a Hip build of Trilinos using Kokkos as external TPL (rocm/5.6.1 + gcc/8.5.0) I'm seeing these types of errors, some  apparent gcc/8 incompatibility

```
08:21:23 In file included from <built-in>:1:
08:21:23 In file included from /opt/rocm-5.6.1/llvm/lib/clang/16.0.0/include/__clang_hip_runtime_wrapper.h:50:
08:21:23 In file included from /opt/rocm-5.6.1/llvm/lib/clang/16.0.0/include/cuda_wrappers/cmath:27:
08:21:23 In file included from /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/cmath:1892:
08:21:23 In file included from /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/specfun.h:45:
08:21:23 In file included from /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_algobase.h:64:
08:21:23 In file included from /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/stl_pair.h:59:
08:21:23 In file included from /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/bits/move.h:55:
08:21:23 /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/type_traits:1000:56: error: incomplete type 'Kokkos::Impl::UnsupportedKokkosArrayLayout' used in type trait expression
08:21:23       : public __bool_constant<__is_assignable(_Tp, _Up)>
08:21:23                                                        ^
08:21:23 /usr/lib/gcc/x86_64-redhat-linux/8/../../../../include/c++/8/type_traits:2799:43: note: in instantiation of template class 'std::is_assignable<Kokkos::mdspan<double, Kokkos::extents<unsigned long, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615>, Kokkos::Experimental::layout_right_padded<>>, Kokkos::Impl::UnsupportedKokkosArrayLayout>' requested here
08:21:23   inline constexpr bool is_assignable_v = is_assignable<_Tp, _Up>::value;
08:21:23                                           ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Build/install_kokkos/include/Kokkos_View.hpp:1740:46: note: in instantiation of variable template specialization 'std::is_assignable_v<Kokkos::mdspan<double, Kokkos::extents<unsigned long, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615>, Kokkos::Experimental::layout_right_padded<>>, Kokkos::Impl::UnsupportedKokkosArrayLayout>' requested here
08:21:23             typename = std::enable_if_t<std::is_assignable_v<
08:21:23                                              ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Build/install_kokkos/include/Kokkos_View.hpp:1744:36: note: in instantiation of default argument for 'operator mdspan<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2, type-parameter-0-3><double, Kokkos::extents<unsigned long, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615>, Kokkos::Experimental::layout_right_padded<>, Kokkos::default_accessor<double>>' required here
08:21:23   KOKKOS_INLINE_FUNCTION constexpr operator mdspan<
08:21:23                                    ^~~~~~~~~~~~~~~~
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Trilinos/packages/sacado/src/KokkosExp_View_Fad.hpp:408:67: note: while substituting deduced template arguments into function template 'operator mdspan<type-parameter-0-0, type-parameter-0-1, type-parameter-0-2, type-parameter-0-3>' [with OtherElementType = double, OtherExtents = Kokkos::extents<unsigned long, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615, 18446744073709551615>, OtherLayoutPolicy = Kokkos::Experimental::layout_right_padded<>, OtherAccessor = Kokkos::default_accessor<double>, $4 = (no value)]
08:21:23   typename PODViewDeepCopyType< View<DT,DP...> >::type dst_array( dst );
08:21:23                                                                   ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Trilinos/packages/sacado/src/KokkosExp_View_Fad.hpp:408:56: note: while substituting deduced template arguments into function template 'View' [with U = (no value)]
08:21:23   typename PODViewDeepCopyType< View<DT,DP...> >::type dst_array( dst );
08:21:23                                                        ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Trilinos/packages/sacado/src/KokkosExp_View_Fad.hpp:435:11: note: in instantiation of function template specialization 'Kokkos::deep_copy<Kokkos::Serial, Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ****, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial, Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ****, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks>' requested here
08:21:23   Kokkos::deep_copy(exec_space(), dst, src);
08:21:23           ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Trilinos/packages/sacado/test/performance/advection/common.hpp:122:11: note: in instantiation of function template specialization 'Kokkos::deep_copy<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ****, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial, Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ****, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::Experimental::EmptyViewHooks>' requested here
08:21:23   Kokkos::deep_copy( v1, v1_h );
08:21:23           ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Trilinos/packages/sacado/test/performance/advection/advection_hierarchical_dfad.cpp:140:3: note: in instantiation of function template specialization 'init_fad<Kokkos::View<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ****, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial>, Kokkos::View<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ***, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial>, Kokkos::View<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> ***, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial>, Kokkos::View<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> **, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial>, Kokkos::View<Sacado::Fad::Exp::GeneralFad<Sacado::Fad::Exp::DynamicStorage<double>> **, Kokkos::LayoutContiguous<Kokkos::LayoutRight>, Kokkos::Serial>>' requested here
08:21:23   init_fad(wgb, wbs, flux, src, residual);
08:21:23   ^
08:21:23 /home/jenkins/caraway-new/workspace/KokkosEco_AsTPLs_Trilinos_Caraway_Rocm561_Hip_Serial_MI210_rocmtpls/Build/install_kokkos/include/Kokkos_View.hpp:378:8: note: forward declaration of 'Kokkos::Impl::UnsupportedKokkosArrayLayout'
08:21:23 struct UnsupportedKokkosArrayLayout;
```

**Comment by crtrott** [comment]
In both cases we need to have a new enough GCC underneath. On Blake you can now also load a newer GCC module at the same time as you load icpx, and then add `--gcc-toolchain=${GCC_ROOT}` to the CMAKE_CXX_FLAGS. This might also work for the HIP build.

**Comment by ndellingwood** [comment]
Working on #7103 regarding the Trilinos include issues; compiling with Cuda enabled I'm seeing lots of these types of warnings:
```
/home/ndellin/Trilinos/kokkos/core/src/../../tpls/mdspan/include/mdspan/../experimental/__p2630_bits/submdspan_mapping.hpp(134): warning: invalid narrowing conversion from "int" to "size_t"
```

**Comment by masterleinad** [comment]
> Working on #7103 regarding the Trilinos include issues; compiling with Cuda enabled I'm seeing lots of these types of warnings:

Also see https://github.com/kokkos/mdspan/pull/350.

**Comment by ndellingwood** [comment]
A couple separate items now addressed:
* Failures with icpx with c++20 in https://github.com/kokkos/kokkos/issues/7075#issue-2358493922 resolved by using a newer version of gcc in the toolchain; this did not resolve the issues with icpc, and we opted to not test icpc with c++20 with mdspan enabled
* The Trilinos failures addressed by #7103
* The many warnings in Cuda builds will be addressed by https://github.com/kokkos/mdspan/pull/350

**Comment by dalg24** [comment]
I expect this was resolved in #7118

**Comment by dalg24** [comment]
> I expect this was resolved in #7118

Not that's not it.  It looks more like https://github.com/kokkos/kokkos/pull/7103
@ndellingwood please confirm this is resolved

**Comment by ndellingwood** [comment]
Yes, this is resolved
